### DISCUSSION TITLE
Cell metadata format

### DISCUSSION THREAD

**Discussion by akhmerov** [executablebooks/meta#115] [Ideas] [answered]
#243 introduces a way to control how the cell output is rendered by the cell based on:

- `myst` metadata dictionary
- `remove_stdout`/`remove_stderr` cell tags

First of all, cell metadata seems to be a great place to control and  customize output rendering. I'm still uncertain whether figures wouldn't be better off as a container around a complete cell, similar to latex, but that is an aside.

Still, several follow-up questions come to mind:

- Wouldn't cell metadata be a nice language-agnostic implementation of glue? Previously there was no mechanism to implement this, but now it is in place.
- Is `myst` metadata namespace the best option? After all, it really has no relation to the myst document format, and is exclusively about how sphinx should render cell outputs. `sphinx`/`render` or any combination of those would seem more logical.
- What is the motivation for using both a metadata dictionary and tags? It seems more complex than storing everything in a single dictionary.

**Comment by chrisjsewell** [comment, marked as answer]
Cheers 😄 

> Wouldn't cell metadata be a nice language-agnostic implementation of glue?

Yeh "language-agnostic" is certainly a big plus. The two main cons of cell metadata I see are:

1. The id tag would not always be easily visible in notebook environments, in jupyterlab you can have the sidebar viewer 👍, but in the regular notebook, and e.g.in VS Code, its a pain/impossible to access.
2. Dealing with multiple outputs of a cell; its less specific than "wrapping" a particular object in the glue function and setting an id

> Is myst metadata namespace the best option? 

Well I made it up at about 5 hours ago, so who knows lol. It in the code as a global variable, so easy to change. Whatever, people are happy with really (could also be a sphinx config variable)

> What is the motivation for using both a metadata dictionary and tags? 

As with (1) tags are easy to manipulate in the jupyter environment, and more concise and simpler for more "average" users than nested keys. But obviously those are required if dealing with more complex formatting requirements, like the image/figure example.
Certainly though, when writing in the "MyST Markdown Notebooks", it makes it so much easier to manipulate this metadata, and that is less of an issue there.

**Reply by akhmerov** [reply]
> The id tag would not always be easily visible in notebook environments

The alternatives aren't mutually exclusive though: both mechanisms could be in place.

> Dealing with multiple outputs of a cell

I think in practice this would never be a limitation, since the author also controls the cell contents.

> Is myst metadata namespace the best option?
> 
> > Well I made it up at about 5 hours ago, so who knows lol.

Didn't you also mint a release about 4 hours ago..?

**Comment by chrisjsewell** [comment]
> Didn't you also mint a release about 4 hours ago..?

Yep, I don't mess around! The namespace is now configurable and set to `render` in v0.9.1:
https://myst-nb.readthedocs.io/en/latest/use/formatting_outputs.html#images

**Reply by chrisjsewell** [reply]
Thats the ethos in our design philosophy: https://github.com/executablebooks/.github/blob/master/CONTRIBUTING.md#design-philosophy, "release early and often"
Still waiting on that jupyter-sphinx v0.3 😜
